mlir/test/Target/LLVMIR/genx.mlir

```diff
@@ -0,0 +1,31 @@
+// RUN: mlir-translate -mlir-to-llvmir %s | FileCheck %s
+
+llvm.func @genx_special_regs() -> i64 {
+  // CHECK-LABEL: genx_special_regs
+  // CHECK: call i64 @get_local_id(i32 0)
```

Comment:
@pengtu `createDeviceFunctionCall` hardcodes the return type to be `i64`, can `get_local_id` return `i32`? If yes, will change `createDeviceFunctionCall` to generate the same return type as `genx.workitem.id.x`. If no, will change the codegen to add cast operation.

Reply:
Looks like the SPIRV-LLVM translator has support for these builtins returning either i32 or i64:
```
test/builtin-globals.ll:; CHECK-FUNCTION-FORMAT-OCL-REV: declare spir_func i64 @_Z12get_group_idj(i32)
test/transcoding/OpGroupAsyncCopy.ll:  %call15 = call spir_func i32 @_Z12get_group_idj(i32 0)

test/transcoding/OpGroupAsyncCopy.ll:declare spir_func i32 @_Z13get_global_idj(i32) #1
test/extensions/INTEL/SPV_INTEL_function_pointers/non-uniform-function-pointer.ll:declare spir_func i64 @_Z13get_global_idj(i32) #2
```
So we could just generate the one that matches the GENX operation return type and avoid the type conversion if we wanted to.

Reply:
I am not sure.  In all the SYCL IRs that I have seen, they are defaulting to 64 bit addressing machine models and using 64 bit return values for the get_group_id() and etc.  The createDeviceFunctionCall is copied from ROCDL, so AMD is also defaulting to 64 bit.  My current Triton-dse prototype also uses SPIRV dialect's 64 bit memory model with 64 bit and a cast.  32 bit version may work but I haven't tested it.  I would say that we default to 64 and add a cast for now.  We can revisit this later.

Reply:
/// Creates a pass that lowers GPU dialect operations to GENX counterparts. The
/// index bitwidth used for the lowering of the device side index computations
/// is configurable.
std::unique_ptr<OperationPass<gpu::GPUModuleOp>>
createLowerGpuOpsToGENXOpsPass(
    unsigned indexBitwidth = kDeriveIndexBitwidthFromDataLayout);
} // namespace mlir

BTW, the MLIR's indexBitwidth is coming from the this function parameter.  We can set the device function return type to match this if we don't default to 64 bit.

Reply:
A correction, AMD actually uses 32 bit intrinsics for the indices and 64 bit device functions for the work_group_size() etc. NVVM also uses 32 bit for indices.

We generate device functions for both indices and work_group_size().  I think that we need to have two versions of the device function generators, one returns 32 bit  and one returns 64 bit result types. 

Thoughts?

Reply:
https://github.com/pengtu/llvm-project/blob/genx/mlir/lib/Conversion/GPUCommon/IndexIntrinsicsOpLowering.h#L23

The GPUIndexIntrinsicLowering function assumes the XOp, YOp, ZOp aways return 32 bit value.  If we change the device function creation to return int32, it will work.

Reply:
Sure, will work on it.

Reply:
Should we call `llvm.genx.GenISA.getLocalID.X` (https://github.com/intel-innersource/drivers.gpu.unified/blob/comp/igc/Source/IGC/GenISAIntrinsics/Intrinsic_definitions.py#L1219) instead? As it returns `i16`, we would also need a cast.

Reply:
I think before doing anything we should have an issue to document the design. We should discuss ion the issue how to lower the Triton operation and once we have consensus go and implement the feature.